File: jar/src/main/java/jline/api/mc/Dtmc_transient.java
/**
 * @file Transient distribution of a discrete-time Markov chain
 *
 * @since LINE 3.0
 */
package jline.api.mc;

import jline.util.matrix.Matrix;

public final class Dtmc_transient {
    private Dtmc_transient() {}

    /**
     * Transient distribution of a DTMC, i.e. the rows pi(k) = pi0*P^k for
     * k = 0,...,steps. Twin of the MATLAB dtmc_transient and of the Python
     * api.mc.dtmc_transient.
     *
     * @param P     transition matrix
     * @param pi0   initial distribution, uniform when null
     * @param steps number of steps
     * @return matrix with one row per step, from step 0 to step steps
     */
    public static Matrix dtmc_transient(Matrix P, Matrix pi0, int steps) {
        int n = P.getNumRows();
        if (steps < 0) {
            throw new IllegalArgumentException("The number of steps must be non-negative.");
        }
        Matrix pik = new Matrix(1, n);
        for (int i = 0; i < n; i++) {
            pik.set(0, i, (pi0 == null || pi0.length() != n) ? 1.0 / n : pi0.get(i));
        }
        Matrix pit = new Matrix(steps + 1, n);
        for (int k = 0; k <= steps; k++) {
            for (int i = 0; i < n; i++) {
                pit.set(k, i, pik.get(0, i));
            }
            pik = pik.mult(P);
        }
        return pit;
    }
}

File: matlab/src/api/mc/dtmc_transient.m
function pi_t = dtmc_transient(P, pi0, steps)
% PI_T = DTMC_TRANSIENT(P, PI0, STEPS)
%
% Transient distribution of a DTMC with transition matrix P, i.e. the rows
% PI(k) = PI0*P^k for k = 0,...,STEPS. Twin of the Python api.mc.dtmc_transient.
%
% Copyright (c) 2012-2026, Imperial College London
% All rights reserved.

n = size(P,1);
if nargin < 2 || isempty(pi0)
    pi0 = ones(1,n)/n;
end
pi0 = reshape(pi0, 1, n);
if nargin < 3 || isempty(steps)
    steps = 1;
end
if steps < 0 || steps ~= round(steps)
    line_error(mfilename,'The number of steps must be a non-negative integer.');
end

pi_t = zeros(steps+1, n);
pik = pi0;
pi_t(1,:) = pik;
for k = 1:steps
    pik = pik * P;
    pi_t(k+1,:) = pik;
end
end
